strawberryfields/parameters.py

```diff
@@ -19,335 +19,334 @@
+            # apply func elementwise, recursively, on the args
+            return np.array([wrapper(*k) for k in zip(*args)])
+        return func(*args)
+    return wrapper
 
-from .program_utils import (RegRef, RegRefTransform)
+# namespace of mathematical functions for manipulating Parameters
+parfuncs = types.SimpleNamespace(**{name: wrap_mathfunc(getattr(sf, name)) for name in dir(sf) if name[0] != '_'})
 
 
-try:
-    import tensorflow as tf
-    _tf_classes = (tf.Tensor, tf.Variable)
-except ImportError:
-    tf = mock.MagicMock()
-    _tf_classes = tuple()
 
+class ParameterError(RuntimeError):
+    """Exception raised when the Parameter classes encounter an illegal operation.
+
+    E.g., trying to use a measurement result before it is available.
+    """
+
+
+def is_object_array(p):
+    """Returns True iff p is an object array."""
+    return isinstance(p, np.ndarray) and p.dtype == object
 
-def _unwrap(params):
-    """Unwrap a parameter sequence.
+
+def par_evaluate(params):
+    """Evaluate a parameter sequence.
+
+    Any parameters descending from sympy.Basic are evaluated, others are returned as is.
+    Evaluation means that free and measured parameters are replaced by their numeric values.
+    Numpy arrays are evaluated elementwise.
+
+    Alternatively, evaluates a single parameter and returns its value.
 
     Args:
-      params (Sequence[Parameter]): parameters to unwrap
+        params (Sequence[Any]): parameters to evaluate
 
     Returns:
-      tuple[Number, array, Tensor, Variable]: unwrapped Parameters
+        list[Any]: evaluated parameters
+    """
+    scalar = False
+    if not isinstance(params, Sequence):
+        scalar = True
+        params = [params]
+
+    def do_evaluate(p):
+        if is_object_array(p):
+            return np.array([do_evaluate(k) for k in p])
+
+        if not par_is_symbolic(p):
+            return p
+
+        #p = p.evalf()  # TODO sympy 1.4 has some bugs which prevent using this for now, workaround below
+        if 0:
+            # build the substitution dict for binding free and measured params
+            d = {}
+            for k in p.atoms(MeasuredParameter, FreeParameter):
+                d[k] = k._eval_evalf(None)
+            p = p.evalf(subs=d)
+            # TODO evalf cannot handle substitutions with arbitrary objects even if p is atomic, hence substituting e.g. numpy arrays or TF classes causes sympy to raise an error
+            # TODO the float() conversion below prevents symbolic params from being passed through, maybe the backend should do float() conversion?
+            if not p.is_real:
+                return complex(p)
+            if p.is_integer:
+                return int(p)
+            return float(p)
+
+        # using lambdify we can also substitute np.ndarrays and tf.Tensors for the atoms
+        atoms = list(p.atoms(MeasuredParameter, FreeParameter))
+        func = sympy.lambdify(atoms, p, 'numpy')
```

Comment:
How does this work if `p` contains tensorflow variables?

Reply:
Or is this just a recognition that when we are evaluating, the output should be a numpy object?

Reply:
Good question. Roughly speaking `lambdify` converts the symbolic expression `p` into a string of python code, and then evaluates that code in a specific namespace (which provides definitions for the various functions) to create a lambda function that substitutes the passed argument values for `atoms`.

Here that namespace is `numpy`. This means that you probably cannot apply functions such as `sin` (which is translated to `np.sin`) to TF objects, but basic arithmetic operations should work. 

lambdify could also use `tensorflow` as the namespace if needed, but I'm not sure what would be the best way to choose the namespace.

https://docs.sympy.org/latest/modules/utilities/lambdify.html